sdlib/d/gc/bin.d

```diff
@@ -31,21 +31,27 @@ struct Bin {
 		auto size = binInfos[sizeClass].itemSize;
 
 		Extent* slab;
-		uint index;
+		uint gotAllocs;
 
 		{
 			mutex.lock();
 			scope(exit) mutex.unlock();
 
 			slab = (cast(Bin*) &this).getSlab(arena, emap, sizeClass);
 			if (slab is null) {
-				return null;
+				return 0;
 			}
 
-			index = slab.allocate();
+			// Get as many of the requested slots as we can from the current slab:
+			import d.gc.util;
+			gotAllocs = min(cast(uint) allocs.length, slab.freeSlots);
+			foreach (i; 0 .. gotAllocs) {
+				auto index = slab.allocate();
```

Comment:
Looks like to me you need a batch alloc there too instead of searching for the unset bits again and again and updating the availabel slot count every time.

Hint: this is how jemalloc does it, and we use a similar technique: https://github.com/jemalloc/jemalloc/blob/5832ef658975d5f2da2bdfddf55712d9fa343e30/src/arena.c#L245-L291